src/main/java/uk/gov/hmcts/reform/userprofileapi/service/UserProfileUpdator.java

```diff
@@ -28,42 +41,116 @@
 
         HttpStatus status = HttpStatus.OK;
         UserProfile userProfile = null;
         if (!isUserIdValid(userId, false)) {
-            persistAudit(HttpStatus.NOT_FOUND, null);
+            persistAudit(HttpStatus.NOT_FOUND, null, ResponseSource.SYNC);
             throw new ResourceNotFoundException("userId provided is malformed");
         }
 
         Optional<UserProfile> userProfileOptional = userProfileRepository.findByIdamId(userId);
         userProfile =  userProfileOptional.orElse(null);
 
         if (userProfile == null) {
-            persistAudit(HttpStatus.NOT_FOUND, null);
+            persistAudit(HttpStatus.NOT_FOUND, null, ResponseSource.SYNC);
             throw new ResourceNotFoundException("could not find user profile for userId: " + userId);
         } else if (!isUpdateUserProfileRequestValid(updateUserProfileData)) {
-            persistAudit(HttpStatus.BAD_REQUEST, null);
+            persistAudit(HttpStatus.BAD_REQUEST, null,ResponseSource.SYNC);
             throw new RequiredFieldMissingException("Update user profile request is not valid for userId: " + userId);
         } else if (!isSameAsExistingUserProfile(updateUserProfileData, userProfile)) {
+
             userProfile.setEmail(updateUserProfileData.getEmail().trim());
             userProfile.setFirstName(updateUserProfileData.getFirstName().trim());
             userProfile.setLastName(updateUserProfileData.getLastName().trim());
             userProfile.setStatus(IdamStatus.valueOf(updateUserProfileData.getIdamStatus().toUpperCase()));
-            try {
-                userProfile = userProfileRepository.save(userProfile);
-            } catch (Exception ex) {
-                status = HttpStatus.INTERNAL_SERVER_ERROR;
-            }
         }
-        persistAudit(status, userProfile);
+
+        try {
+            userProfile = userProfileRepository.save(userProfile);
+        } catch (Exception ex) {
+            status = HttpStatus.INTERNAL_SERVER_ERROR;
+        }
+        persistAudit(status, userProfile, ResponseSource.SYNC);
         return userProfile;
     }
 
-    public void persistAudit(HttpStatus idamStatus, UserProfile userProfile) {
-        Audit audit = new Audit(idamStatus.value(), resolveStatusAndReturnMessage(idamStatus), ResponseSource.SYNC, userProfile);
+    @Override
+    public UserProfileRolesResponse updateRoles(UpdateUserProfileData profileData, String userId) {
+        UserProfile userProfile = null;
+        HttpStatus httpStatus = null;
+        UserProfileRolesResponse userProfileRolesResponse =  new UserProfileRolesResponse();
+        userProfile = validateUserStatus(userId);
+        if (!CollectionUtils.isEmpty(profileData.getRolesAdd())) {
+            log.info("Add idam roles for userId :" + userId);
+            AddRoleResponse addRoleResponse = new AddRoleResponse();
+            try (Response response = idamClient.addUserRoles(profileData.getRolesAdd(), userId)) {
```

Comment:
idamClient.addUserRoles can be reused directly from SidamService which is already created. If in future if addUserRoles functionality of SIDAM is required, then we can reuse from sidamService.